async/async_request_processor.go

```diff
@@ -0,0 +1,191 @@
+func NewAsyncRequestWorkerWithFunc[T any, R any](
+	bufferSize int,
+	retryConfig *retry.RetryConfig,
+	processFn func(ctx context.Context, req Request[T]) (R, error),
+) *AsyncRequestProcessor[T, R] {
+	processor := FunctionProcessor[T, R]{ProcessFn: processFn}
+	return NewAsyncRequstProcessor(bufferSize, processor, retryConfig)
+}
+
+// Start begins the worker's processing loop in a separate goroutine
+func (w *AsyncRequestProcessor[T, R]) Start() {
+	w.wg.Add(1)
+	go w.processLoop()
+}
+
+// Stop gracefully shuts down the worker after processing remaining requests
+func (w *AsyncRequestProcessor[T, R]) Stop() {
+	w.cancel()
+	w.wg.Wait()
+	close(w.responseChan)
+}
+
+// Submit sends a new request to the worker
+// Returns false if the worker is unable to accept the request
+func (w *AsyncRequestProcessor[T, R]) Submit(req Request[T]) bool {
+	select {
+	case <-w.ctx.Done():
+		return false
+	case w.requestChan <- req:
+		return true
+	default:
+		// Channel is full
+		return false
+	}
+}
+
+// Responses returns the channel for receiving responses
+func (w *AsyncRequestProcessor[T, R]) Responses() <-chan Response[R] {
+	return w.responseChan
+}
+
+// processLoop is the main worker routine that processes requests synchronously
+func (w *AsyncRequestProcessor[T, R]) processLoop() {
+	defer w.wg.Done()
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			// Process remaining items in the channel before exiting
+			for {
+				select {
+				case req := <-w.requestChan:
+					w.processRequest(req)
+				default:
+					return
+				}
+			}
+
+		case req := <-w.requestChan:
+			w.processRequest(req)
+		}
+	}
+}
+
+// processRequest handles processing a single request with retry logic
+func (w *AsyncRequestProcessor[T, R]) processRequest(req Request[T]) {
+	startTime := time.Now()
+
+	var responseData R
+	var err error
+
+	// If no retry config is set, process the request directly
+	if w.retryConfig == nil {
+		responseData, err = w.process(req)
+	} else {
+		// Retry logic
+		err = retry.RetryWithBackoff(w.ctx, *w.retryConfig, func(ctx context.Context) error {
+			responseData, err = w.process(req)
+			return err
+		})
+	}
+
+	duration := time.Since(startTime)
+
+	// Send the response back through the response channel
+	select {
+	case w.responseChan <- Response[R]{
+		RequestID: req.ID,
+		Data:      responseData,
+		Error:     err,
+		Duration:  duration,
+	}:
+	case <-w.ctx.Done():
+		// Worker is shutting down, don't try to send results
+	}
+}
+
+func (w *AsyncRequestProcessor[T, R]) process(req Request[T]) (R, error) {
+	// Create a context for this specific request that inherits from the worker context
+	reqCtx, cancel := context.WithTimeout(w.ctx, w.retryConfig.MaxDuration)
+
+	// Process the request using the custom processor
+	responseData, err := w.processor.Process(reqCtx, req)
+	cancel() // Always cancel the request context
+
+	if err == nil {
+		return responseData, nil
+	}
+
+	return responseData, err
+}
```

Comment:
This file looks familiar, is it taken from the async processing in polaris?

Reply:
No, completely new